### ISSUE TITLE
Plot.y_range / Plot.x_range initial start/end values not set

### ISSUE THREAD

**Issue by dhirschfeld** [bokeh/bokeh#4371]
After a Plot is created the start/end values of the ranges aren't set on the Python side so AFAICS there is no way to know the existing axis limits.

Example notebook demoing the problem is available at:
https://anaconda.org/dhirschfeld/rangenotsetbug/notebook

**Comment by dhirschfeld** [comment]
Nice to haves:
 * Have the repr of a Range object show the start/end values for convenience
 * Allow setting the range with a 2-element iterable instead of having to set the start/end  properties individually

**Comment by dhirschfeld** [comment]
I confirm that it's still present on master - `0.12.0dev9-43-g560acd8`

**Comment by canavandl** [comment]
@dhirschfeld 

It's possible to set the range as a 2-element tuple either by setting the range bounds inside the `figure` initializer method or by resetting the x/y_range properties on the plot via:

```python
import bokeh.plotting as plt
from bokeh.models import Range1d

desired_range = (0,10)

#option 1
fig = plt.figure(x_range=desired_range)
#option 2
fig = plt.figure(x_range=Range1d(*desired_range)) #note the splatting
#option 3
fig = plt.figure()
fig.x_range = Range1d(*desired_range) #there are multiple range types
```

However, it's not possible to read the computed (based plotted data) range bounds because the range bound solver exists on the BokehJS client side and not in python.

**Comment by bryevdv** [comment]
It's possible we could find a way to report this, but we would need to design and create a new facility for consistently handing "client-write-only" properties.

**Comment by dhirschfeld** [comment]
@canavandl - it's certainly possible to set the range by constructing a `Range` object but it would be nice to be able to do it directly without having to box the min/max values in an intermediary object. If it was a simple property I'd implement it like:
```python
@x_range.setter
def x_range(self, value):
    if not isinstance(value, Range):
        # create range of the same class as existing
        value = self.x_range.__class__(*value)
    self.x_range = value
```
...but I see the implementation is more complicated than that so I'm not sure that it can be implemented so simply.

----

Is it really not possible to know on the Python side what the extent of your axes is? What if I wanted to put an annotation in the middle of the plot - e.g.
```
fig.text((xmax - xmin)/2, (ymax - ymin)/2, 'middle')
```
...or more generally for any location wrt axes coordiantes - how can I do that without knowing the axes extents?

**Comment by canavandl** [comment]
Annotations support setting renderer locations either in screen units (related to the pixel distance from the origin of the plot) or data units (which uses the mapper that is used to calculate the plot ranges based on the input data). 

For example, adding a span along the vertical center of the plot would look like:

```python
p = figure(plot_width=600, plot_height=600)

# '300' = plot_height/2
span = Span(location=300, location_units='screen', dimension='width')

p.renderers.append(span)
```

Apart from BoxAnnotations in 0.11.1, several more annotation types (Labels, Arrows, etc) are currently in master and due to be released in 0.12.0 in the next couple of weeks. We're also going to cut some dev builds in the next days, if you want to try it sooner.

(I also think adding a `setter` on ranges to accept 2-element tuples could be a nice convenience method.)

**Comment by dhirschfeld** [comment]
@canavandl - once you've got titles and axes labels it would seem unlikely that the center of the axes would coincide with the center of the canvas?

My specific use case is that I want to put an annotation at the top/middle of the axes. Without being able to query the extent of the axes how can I position the text label at the top?

**Comment by bryevdv** [comment]
we do have "client only" properties now, so it is feasible to add `computed_start`, `computed_end` or similar to `DataRange1d` now.

**Comment by bryevdv** [comment]
This was changed at some point so that date range start/end *do* get reported back, as adding this to e.g. the sliders demo will show:
```python
def cb(attr, old, new):
    print(attr, old, new)

plot.x_range.on_change('start', cb)
plot.x_range.on_change('end', cb)
```
However the first initial start/end do not trigger a callback, so my inclination is just to make sure that they do, and not add any new properties.

**Comment by VineethKanaparthi** [comment]
Do we have any method to extract computed x_range and y_range?

**Comment by bryevdv** [comment]
@manuvazquez This issue is still open, if there was any news, it would be here. 

For the other exception: you cannot share bokeh objects between different documents/sessions, as the exception states, for a variety of reasons, among them security/privacy between user sessions. This is not going to change.

**Comment by manuvazquez** [comment]
Ok, sorry for the noise (and thank you for the heads up!!)

**Comment by mattpap** [comment]
This was fixed in PR #11777. Sample output when `document_ready` and `change` events are set up:
```
022-08-06 23:57:43,402 Sending pull-doc-reply from session 'M5uxhB62oGRPs3ycm383E4WDdwJvFUSqKQFIEvyqMqX6'
x-range start nan 0.8999999999999999
x-range end nan 3.1
y-range start nan 9
y-range end nan 31
document_ready x-range 0.8999999999999999 3.1
document_ready y-range 9 31
```
Further discussion and issues regarding data range initialization can be found in issues #11798 and #11718.
